base_custom_info/models/custom_info.py

```diff
@@ -85,7 +119,126 @@ class CustomInfoValue(models.Model):
         required=True,
         string='Property')
     name = fields.Char(related='property_id.name', readonly=True)
-    value = fields.Char(translate=True, index=True)
+    field_type = fields.Selection(related="property_id.field_type")
+    field_name = fields.Char(
+        compute="_compute_field_name",
+        help="Technical name of the field where the value is stored.",
+    )
+    value = fields.Char(
+        compute="_compute_value",
+        inverse="_inverse_value",
+        help="Value, always converted to/from the typed field.",
+    )
+    value_str = fields.Char(
+        string="Text value",
+        translate=True,
+        index=True,
+    )
+    value_int = fields.Integer(
+        string="Whole number value",
+        index=True,
+    )
+    value_float = fields.Float(
+        string="Decimal number value",
+        index=True,
+    )
+    value_bool = fields.Boolean(
+        string="Yes/No value",
+        index=True,
+    )
+    value_id = fields.Many2one(
+        comodel_name="custom.info.option",
+        string="Selection value",
+        ondelete="cascade",
+        domain="[('property_ids', 'in', [property_id])]",
+    )
+
+    @api.multi
+    def check_access_rule(self, operation):
+        """You access a value if you access its property and owner record."""
+        for s in self:
+            s.property_id.check_access_rule(operation)
+            record = self.env[s.model].browse(s.res_id)
+            record.check_access_rights(operation)
+            record.check_access_rule(operation)
+        return super(CustomInfoValue, self).check_access_rule(operation)
+
+    @api.multi
+    @api.depends("property_id.field_type")
+    def _compute_field_name(self):
+        """Get the technical name where the real typed value is stored."""
+        for s in self:
+            s.field_name = "value_{!s}".format(s.property_id.field_type)
+
+    @api.multi
+    @api.depends("property_id.field_type", "field_name", "value_str",
+                 "value_int", "value_float", "value_bool", "value_id")
+    def _compute_value(self):
+        """Get the value as a string, from the original field."""
+        for s in self:
+            try:
+                s.value = s[s.field_name]
```

Comment:
However @Yajo is actually being more Pythonic + efficient here IMO due to utilization of the else in the try block. In order to implement that, a new conditional would need to be created to eval vs. just going one way or the other based on a bubbled exception.